matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomFromLocalRoomTask.kt

```diff
@@ -17,133 +17,109 @@
-import org.matrix.android.sdk.internal.util.time.Clock
-import java.util.UUID
 import java.util.concurrent.TimeUnit
 import javax.inject.Inject
 
 /**
  * Create a room on the server from a local room.
  * The configuration of the local room will be use to configure the new room.
  * The potential local room members will also be invited to this new room.
- *
- * A local tombstone event will be created to indicate that the local room has been replacing by the new one.
  */
 internal interface CreateRoomFromLocalRoomTask : Task<CreateRoomFromLocalRoomTask.Params, String> {
     data class Params(val localRoomId: String)
 }
 
 internal class DefaultCreateRoomFromLocalRoomTask @Inject constructor(
-        @UserId private val userId: String,
         @SessionDatabase private val monarchy: Monarchy,
         private val createRoomTask: CreateRoomTask,
-        private val stateEventDataSource: StateEventDataSource,
-        private val clock: Clock,
+        private val roomSummaryDataSource: RoomSummaryDataSource,
 ) : CreateRoomFromLocalRoomTask {
 
     private val realmConfiguration
         get() = monarchy.realmConfiguration
 
     override suspend fun execute(params: CreateRoomFromLocalRoomTask.Params): String {
-        val replacementRoomId = stateEventDataSource.getStateEvent(params.localRoomId, EventType.STATE_ROOM_TOMBSTONE, QueryStringValue.IsEmpty)
-                ?.content.toModel<RoomTombstoneContent>()
-                ?.replacementRoomId
+        val localRoomSummary = roomSummaryDataSource.getLocalRoomSummary(params.localRoomId)
+                ?.takeIf { it.createRoomParams != null && it.roomSummary != null }
+                ?: error("Invalid LocalRoomSummary for ${params.localRoomId}")
 
-        if (replacementRoomId != null) {
-            return replacementRoomId
+        // If a room has already been created for the given local room, return the existing roomId
+        if (localRoomSummary.replacementRoomId != null) {
+            return localRoomSummary.replacementRoomId
         }
 
-        var createRoomParams: CreateRoomParams? = null
-        var isEncrypted = false
-        monarchy.doWithRealm { realm ->
-            realm.where<LocalRoomSummaryEntity>()
-                    .equalTo(LocalRoomSummaryEntityFields.ROOM_ID, params.localRoomId)
-                    .findFirst()
-                    ?.let {
-                        createRoomParams = it.createRoomParams
-                        isEncrypted = it.roomSummaryEntity?.isEncrypted.orFalse()
-                    }
-        }
-        val roomId = createRoomTask.execute(createRoomParams!!)
+        return createRoom(localRoomSummary)
+    }
+
+    private suspend fun createRoom(localRoomSummary: LocalRoomSummary): String {
+        updateCreationState(localRoomSummary.roomId, LocalRoomCreationState.CREATING)
+        val replacementRoomId = runCatching {
+            createRoomTask.execute(localRoomSummary.createRoomParams!!)
```

Comment:
Maybe check if createRoomParams is not null? Not sure it's possible 
And throw a specific error if any?

Reply:
this is checked at line 60, no?

Reply:
I improved the code to avoid the use of `!!` https://github.com/vector-im/element-android/pull/7125/commits/eac74bda0932931b7ec9d53383f813c748d8a12e